File: src/main/java/top/chiloven/lukosbot2/commands/kemono/model/Service.java
package top.chiloven.lukosbot2.commands.kemono.model;

import lombok.Getter;

import java.net.URI;
import java.util.HashMap;
import java.util.Map;
import java.util.regex.Pattern;

/**
 * The service that kemono.cr supports
 */
@Getter
public enum Service {
    PATREON("patreon", "Patreon"),
    FANBOX("fanbox", "Pixiv Fanbox"),
    //DISCORD("discord", "Discord"), (currently not supported)
    FANTIA("fantia", "Fantia"),
    AFDIAN("afdian", "Afdian"),
    BOOSTY("boosty", "Boosty"),
    GUMROAD("gumroad", "Gumroad"),
    SUBSCRIBE_STAR("subscribestar", "SubscribeStar"),
    DL_SITE("dlsite", "DLSite"),
    UNKNOWN("unknown", "Unknown");

    private static final Map<String, Service> SERVICE_MAP = new HashMap<>();

    static {
        for (Service s : values()) {
            SERVICE_MAP.put(s.id, s);
        }
    }

    private final String id;
    private final String name;

    Service(String id, String name) {
        this.id = id;
        this.name = name;
    }

    public static Service getService(String id) {
        return SERVICE_MAP.getOrDefault(id, UNKNOWN);
    }

    /**
     * Parse the post information and the specific service in the service URL.
     *
     * @param uri a URI stored the URL
     * @return a {@link ServiceAndPostId} instance stored with a {@link Service} and a {@link String} service post id
     * @see ServiceAndPostId
     */
    public static ServiceAndPostId parseServicePostUrl(URI uri) {
        String host = uri.getHost() == null ? "" : uri.getHost().toLowerCase();

        try {
            if (host.contains("patreon.com")) {
                String id = Pattern.compile("-(\\d+)(?:[/?#].*)?$").matcher(uri.toString()).group(1);
                return new ServiceAndPostId(PATREON, id);
            }
            if (host.contains("fanbox.cc")) {
                String id = Pattern.compile("/posts/(\\d+)(?:[/?#].*)?$").matcher(uri.toString()).group(1);
                return new ServiceAndPostId(FANBOX, id);
            }
            if (host.contains("fantia.jp")) {
                String id = Pattern.compile("/posts/(\\d+)(?:[/?#].*)?$").matcher(uri.toString()).group(1);
                return new ServiceAndPostId(FANTIA, id);
            }
            if (host.contains("afdian.net")) {
                String id = Pattern.compile("/p/([a-f0-9]+)(?:[/?#].*)?$").matcher(uri.toString()).group(1);
                return new ServiceAndPostId(AFDIAN, id);
            }
            if (host.contains("boosty.to")) {
                String id = Pattern.compile("/posts/([a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12})(?:[/?#].*)?$").matcher(uri.toString()).group(1);
                return new ServiceAndPostId(BOOSTY, id);
            }
            if (host.contains("gumroad.com")) {
                throw new IllegalArgumentException("无法对 Gumroad 的链接进行解析……");
            }
            if (host.contains("subscribestar.com")) {
                String id = Pattern.compile("/posts/(\\d+)(?:[/?#].*)?$").matcher(uri.toString()).group(1);
                return new ServiceAndPostId(SUBSCRIBE_STAR, id);
            }
            if (host.contains("dlsite.com")) {
                String id = Pattern.compile("/product_id/([A-Z0-9]+)\\.html(?:[?#].*)?$").matcher(uri.toString()).group(1);
                return new ServiceAndPostId(DL_SITE, id);
            }
        } catch (Exception e) {
            throw new IllegalArgumentException("错误的平台链接：" + uri);
        }

        throw new IllegalArgumentException("不支持的平台链接: " + uri);
    }

    @Override
    public String toString() {
        return id;
    }

    /**
     * A record with a {@link Service} and a {@link String} service post id.
     *
     * @param service       the service of the post
     * @param servicePostId the service post id
     */
    public record ServiceAndPostId(Service service, String servicePostId) {
    }
}

File: src/main/java/top/chiloven/lukosbot2/commands/impl/kemono/schema/Service.kt
package top.chiloven.lukosbot2.commands.impl.kemono.schema

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonValue
import java.net.URI
import java.util.*

/**
 * Services supported by kemono.
 */
enum class Service(
    @get:JsonValue
    val id: String,
    val siteName: String,
    val abbr: String,
) {

    PATREON("patreon", "Patreon", "p"),
    FANBOX("fanbox", "Pixiv Fanbox", "f"),
    DISCORD("discord", "Discord", "d"),
    FANTIA("fantia", "Fantia", "fa"),
    AFDIAN("afdian", "Afdian", "afd"),
    BOOSTY("boosty", "Boosty", "b"),
    GUMROAD("gumroad", "Gumroad", "g"),
    SUBSCRIBE_STAR("subscribestar", "SubscribeStar", "ss"),
    DL_SITE("dlsite", "DLSite", "dl"),
    UNKNOWN("unknown", "Unknown", "Unknown");

    override fun toString(): String = id

    data class ServiceAndPostId(
        val service: Service,
        val servicePostId: String
    )

    companion object {

        private val byAbbr = entries.associateBy { it.abbr }
        private val byId = entries.associateBy { it.id }

        @JvmStatic
        fun getService(code: String): Service {
            val normalized = code.trim().lowercase(Locale.ROOT)
            return byAbbr[normalized] ?: byId[normalized]
            ?: throw IllegalArgumentException("未知平台：$code")
        }

        @JvmStatic
        @JsonCreator
        fun fromJson(code: String): Service {
            val normalized = code.trim().lowercase(Locale.ROOT)
            return byAbbr[normalized] ?: byId[normalized] ?: UNKNOWN
        }

        fun parseServicePostUrl(uri: URI): ServiceAndPostId {
            val host = (uri.host ?: "").lowercase(Locale.ROOT)
            val raw = uri.toString()

            fun requireId(service: Service, regex: Regex, hint: String): ServiceAndPostId {
                val id = regex.find(raw)?.groupValues?.getOrNull(1)
                    ?: throw IllegalArgumentException("无法从链接中解析 $hint：$uri")
                return ServiceAndPostId(service, id)
            }

            return when {
                "patreon.com" in host -> requireId(
                    PATREON,
                    Regex("(?:/posts/[^/?#]*-|/posts/)(\\d+)(?:[/?#].*)?$"),
                    "Patreon 帖子 ID"
                )

                "fanbox.cc" in host -> requireId(
                    FANBOX,
                    Regex("/posts/(\\d+)(?:[/?#].*)?$"),
                    "Fanbox 帖子 ID"
                )

                "discord.com" in host -> throw IllegalArgumentException(
                    "暂不支持 Discord 内容。"
                )

                "fantia.jp" in host -> requireId(
                    FANTIA,
                    Regex("/posts/(\\d+)(?:[/?#].*)?$"),
                    "Fantia 帖子 ID"
                )

                "afdian.net" in host -> requireId(
                    AFDIAN,
                    Regex("/p/([A-Za-z0-9]+)(?:[/?#].*)?$"),
                    "Afdian 帖子 ID"
                )

                "boosty.to" in host -> requireId(
                    BOOSTY,
                    Regex("/posts/([A-Za-z0-9-]+)(?:[/?#].*)?$"),
                    "Boosty 帖子 ID"
                )

                "subscribestar.com" in host -> requireId(
                    SUBSCRIBE_STAR,
                    Regex("/posts/(\\d+)(?:[/?#].*)?$"),
                    "SubscribeStar 帖子 ID"
                )

                "dlsite.com" in host -> requireId(
                    DL_SITE,
                    Regex("/product_id/([A-Za-z0-9_]+)\\.html(?:[?#].*)?$"),
                    "DLsite 商品 ID"
                )

                "gumroad.com" in host -> throw IllegalArgumentException(
                    "暂不支持从 Gumroad 原站链接解析帖子，请改用 kemono 链接，或提供平台、创作者 ID 和帖子 ID。"
                )

                else -> throw IllegalArgumentException("不支持的平台链接：$uri")
            }
        }

    }

}
